components/script/dom/bindings/record.rs

```diff
@@ -0,0 +1,211 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+//! The `Record` (open-ended dictionary) type.
+
+use crate::dom::bindings::conversions::jsid_to_string;
+use crate::dom::bindings::error::report_pending_exception;
+use crate::dom::bindings::str::{ByteString, DOMString, USVString};
+use indexmap::IndexMap;
+use js::conversions::{ConversionResult, FromJSValConvertible, ToJSValConvertible};
+use js::jsapi::HandleId as RawHandleId;
+use js::jsapi::JSContext;
+use js::jsapi::JS_NewPlainObject;
+use js::jsapi::PropertyDescriptor;
+use js::jsapi::JSITER_HIDDEN;
+use js::jsapi::JSITER_OWNONLY;
+use js::jsapi::JSITER_SYMBOLS;
+use js::jsapi::JSPROP_ENUMERATE;
+use js::jsval::ObjectValue;
+use js::jsval::UndefinedValue;
+use js::rust::wrappers::GetPropertyKeys;
+use js::rust::wrappers::JS_DefineUCProperty2;
+use js::rust::wrappers::JS_GetOwnPropertyDescriptorById;
+use js::rust::wrappers::JS_GetPropertyById;
+use js::rust::wrappers::JS_IdToValue;
+use js::rust::HandleId;
+use js::rust::HandleValue;
+use js::rust::IdVector;
+use js::rust::MutableHandleValue;
+use std::cmp::Eq;
+use std::hash::Hash;
+use std::marker::Sized;
+use std::ops::Deref;
+
+pub trait RecordKey: Eq + Hash + Sized {
+    fn to_utf16_vec(&self) -> Vec<u16>;
+    unsafe fn from_id(cx: *mut JSContext, id: HandleId) -> Option<Self>;
+}
+
+impl RecordKey for DOMString {
+    fn to_utf16_vec(&self) -> Vec<u16> {
+        self.encode_utf16().collect::<Vec<_>>()
+    }
+
+    unsafe fn from_id(cx: *mut JSContext, id: HandleId) -> Option<DOMString> {
+        jsid_to_string(cx, id)
+    }
+}
+
+impl RecordKey for USVString {
+    fn to_utf16_vec(&self) -> Vec<u16> {
+        self.0.encode_utf16().collect::<Vec<_>>()
+    }
+
+    unsafe fn from_id(cx: *mut JSContext, id: HandleId) -> Option<USVString> {
+        rooted!(in(cx) let mut jsid_value = UndefinedValue());
+        let raw_id: RawHandleId = id.into();
+        JS_IdToValue(cx, *raw_id.ptr, jsid_value.handle_mut());
+
+        match USVString::from_jsval(cx, jsid_value.handle(), ()) {
+            Ok(ConversionResult::Success(s)) => Some(s),
+            Ok(ConversionResult::Failure(_)) => None,
+            Err(..) => None,
+        }
+    }
+}
+
+impl RecordKey for ByteString {
+    fn to_utf16_vec(&self) -> Vec<u16> {
+        self.iter().map(|&x| x as u16).collect::<Vec<u16>>()
+    }
+
+    unsafe fn from_id(cx: *mut JSContext, id: HandleId) -> Option<ByteString> {
+        rooted!(in(cx) let mut jsid_value = UndefinedValue());
+        let raw_id: RawHandleId = id.into();
+        JS_IdToValue(cx, *raw_id.ptr, jsid_value.handle_mut());
+
+        match ByteString::from_jsval(cx, jsid_value.handle(), ()) {
```

Comment:
Why does `from_jsval` fails with `Err` and succeeds with `Ok(ConversionResult::Success)` while `Ok(ConversionResult::Failure(_))` never happens?

Reply:
According to https://doc.servo.org/mozjs/conversions/trait.FromJSValConvertible.html#tymethod.from_jsval, Err means that there is a JS exception that needs to be dealt with. I guess the only way that converting a ByteString can fail is if the JS engine throws an exception?

Reply:
Ah okay, that documentation is helpful.

Reply:
So wait, isn't it wrong to just return `None` when the JS engine threw an exception? Isn't that a quite unstable state for it to stay? @jdm Shouldn't that exception immediately make us stop and return back to JS?

Reply:
Returning `Result<>` here will also enable passing failure messages from `ConversionResult::Failure(_)`

Reply:
Sure, but you are inhibiting the actual exception that was thrown by SpiderMonkey itself, which has its own message, I don't think that's the thing to do here.

Reply:
Seems it's not exactly from SM:

https://github.com/servo/servo/blob/ef7b58f0cd02652d2d5c7d626d4a3f9790c080af/components/script/dom/bindings/conversions.rs#L336-L343

Should we use `ConversionResult::Failure` here instead?